cmd/admin-bucket-quota.go

```diff
@@ -0,0 +1,186 @@
+/*
+ * MinIO Client (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+
+	humanize "github.com/dustin/go-humanize"
+	"github.com/fatih/color"
+	"github.com/minio/cli"
+	json "github.com/minio/mc/pkg/colorjson"
+	"github.com/minio/mc/pkg/probe"
+	"github.com/minio/minio/pkg/console"
+	"github.com/minio/minio/pkg/madmin"
+)
+
+var adminQuotaFlags = []cli.Flag{
+	cli.BoolFlag{
+		Name:  "fifo",
+		Usage: "Set fifo quota, allowing automatic deletion of older content",
+	},
+	cli.BoolFlag{
+		Name:  "hard",
+		Usage: "Set a hard quota, disallowing writes after quota is reached",
+	},
+	cli.BoolFlag{
+		Name:  "clear",
+		Usage: "Clears bucket quota configured for bucket",
+	},
+}
+
+// quotaMessage container for content message structure
+type quotaMessage struct {
+	op        string
+	Status    string `json:"status"`
+	Bucket    string `json:"bucket"`
+	Quota     uint64 `json:"quota,omitempty"`
+	QuotaType string `json:"type,omitempty"`
+}
+
+func (q quotaMessage) String() string {
+	if q.op == "set" {
+		return console.Colorize("QuotaMessage",
+			fmt.Sprintf("Successfully set bucket quota of %s with %s type on `%s`", humanize.IBytes(q.Quota), q.QuotaType, q.Bucket))
+	}
+	if q.op == "unset" {
+		return console.Colorize("QuotaMessage",
+			fmt.Sprintf("Successfully cleared bucket quota configured on `%s`", q.Bucket))
+	}
+	return console.Colorize("QuotaInfo",
+		fmt.Sprintf("Bucket `%s` has %s quota of %s", q.Bucket, q.QuotaType, humanize.IBytes(q.Quota)))
+}
+
+func (q quotaMessage) JSON() string {
+	q.Status = "success"
+	jsonMessageBytes, e := json.MarshalIndent(q, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+
+	return string(jsonMessageBytes)
+}
+
+var adminBucketQuotaCmd = cli.Command{
+	Name:   "quota",
+	Usage:  "Manage bucket quota",
+	Action: mainAdminBucketQuota,
+	Before: setGlobalsFromContext,
+	Flags:  append(adminQuotaFlags, globalFlags...),
+	CustomHelpTemplate: `NAME:
+   {{.HelpName}} - {{.Usage}}
+ 
+USAGE:
```

Comment:
I was wondering if the `admin-bucket-quota` command would be more intuitive if we tied the quota size to the type of quota via flags like so,
1. `mc admin bucket quota --hard 100G myminio/my-bucket `
Sets hard quota limit of 100 G on `my-bucket` 

2. `mc admin bucket quota --fifo 75G myminio/my-bucket`
Sets `fifo` quota limit of 75G on `my-bucket` 

3. `mc admin bucket quota --clear myminio/my-bucket`
Clears quotas set on `my-bucket`

4. `mc admin bucket quota myminio/my-bucket`
Gets quota set on `my-bucket`

In this, `--hard` and `--fifo` both require the `size` argument. However, `--clear` is a boolean flag with no argument.
If none of the above flags are present, we return the existing quotas set on the given bucket.

Reply:
That's a good suggestion, changed it @krisis